### PR TITLE
Core: Ignore manager cache on config file changes and Storybook upgrade

### PR DESCRIPTION
**Pull request by ghengeveld**
Issue: #13200 #13291 #13261 

## What I did

Besides looking at the `managerConfig` string, we now also look at modified files in `configDir` (typically `.storybook`) to determine whether or not to use the manager cache. This is achieved by tracking the cache creation date and comparing that to the last modified date on files in `configDir`. This approach doesn't cover every possible way to affect the manager (e.g. a file might be imported from outside the config dir), but it's a step in the right direction.

Additionally, I've made the Storybook version number part of the cache key, so that the cache will get busted whenever Storybook is upgraded.

- `main.js` is ignored because this is what determines the managerConfig in the first place, so any relevant changes there will already bust the cache.
- `preview.js` and `preview-head.html` are ignored because they don't affect the manager.

I've also added some logging to clarify why the cache is not being used.

Note I added some newline logging because otherwise these messages end up getting tucked onto the end of Webpack's progress output, where they're easily overlooked. Unfortunately there's no easy way to fix Webpack's logging, because a custom logger will make log messages go out _after_ the operation they're describing (basically at the next tick), which means if the process hangs for a while, you'll be looking at the _previous_ message and wondering why it takes so long (specifically this "sealing" message at 70% would not show up until it's done sealing, which can take a long time).

## How to test

- Is this testable with Jest or Chromatic screenshots? no
- Does this need a new example in the kitchen sink apps? no
- Does this need an update to the documentation? no

If your answer is yes to any of these, please make sure to include it in your PR.

### CODE REVIEW

File: lib/core-server/src/utils/manager-cache.ts [hunk 1]
```diff
@@ -1,22 +1,59 @@
...
+  // don't use the cache if any file has been modified since the cache was created.
+  const configFiles = await fs.readdir(options.configDir);
+  const cacheCreationDate = new Date(cachedISOTime);
+  try {
+    await Promise.all(
+      configFiles.map(async (file) => {
+        if (ignoredConfigFiles.some((pattern) => pattern.test(file))) return;
+        const filepath = path.join(options.configDir, file);
+        const { mtime: fileModificationDate } = await fs.stat(filepath);
+        if (fileModificationDate > cacheCreationDate) throw filepath;
```
Comment (shilman) [inline lib/core-server/src/utils/manager-cache.ts:42]: How do we guarantee that this is `false` on the very first execution?

Comment (ghengeveld) [reply lib/core-server/src/utils/manager-cache.ts:42]: At first run, there won't be a `cachedISOTime`, so the function will return before reaching this point. See https://github.com/storybookjs/storybook/pull/14993/files#diff-fbd9d4dafd5b84732adcde34aa949c01bcb44e293b5adf8cb512c0bfdcc1cc45R26

**Comment by shilman** [comment]
@ghengeveld There are some conflicts after I merged https://github.com/storybookjs/storybook/pull/15001

Can you please resolve them, and also add this logic to `manager-webpack5` (or refactor to share the logic somehow?). The code currently isn't very DRY, but it's hopefully temporary.

**Comment by ghengeveld** [comment]
> @ghengeveld There are some conflicts after I merged #15001
> 
> Can you please resolve them, and also add this logic to `manager-webpack5` (or refactor to share the logic somehow?). The code currently isn't very DRY, but it's hopefully temporary.

I didn't refactor, but I added it to the webpack5 version now too.

**Comment by shilman** [comment]
@ghengeveld thank you! seems typescript has something to say about it tho:

```
> nx run @storybook/manager-webpack5:prepare 
...
src/index.ts(68,45): error TS2554: Expected 2 arguments, but got 1.
```

**Review by ndelangen — APPROVED** [review]

**Pull request merged**